### PR TITLE
SM/MM: added support for active messaging for bcopy messaging.

### PR DESCRIPTION
**Pull request by alinask**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/868/ for details.

**Comment by alinask** [comment]
@yosefe please review

**Comment by shamisp** [comment]
autobot build

File: src/uct/sm/mm/mm_ep.c [hunk 1]
```diff
@@ -44,6 +53,19 @@ static UCS_CLASS_CLEANUP_FUNC(uct_mm_ep_t)
...
+    uct_mm_remote_seg_t *remote_seg;
+    struct sglib_hashed_uct_mm_remote_seg_t_iterator iter;
+
+    for (remote_seg = sglib_hashed_uct_mm_remote_seg_t_it_init(&iter, self->remote_segments_hash);
+         remote_seg != NULL; remote_seg = sglib_hashed_uct_mm_remote_seg_t_it_next(&iter)) {
+            sglib_hashed_uct_mm_remote_seg_t_delete(self->remote_segments_hash, remote_seg);
+            /* detach the remote proceess's descriptors segment */
+            status = uct_mm_pd_mapper_ops(iface->super.pd)->detach(remote_seg);
+            if (status != UCS_OK) {
+                ucs_warn("Unable to detach shared memory segment of descriptors: %m");
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: need to print status and not %m

File: src/uct/sm/mm/mm_ep.c [hunk 2]
```diff
@@ -89,6 +111,44 @@ ucs_status_t uct_mm_ep_put_bcopy(uct_ep_h tl_ep, uct_pack_callback_t pack_cb,
...
+        /* not in the hash. attach to the memory the mmid refers to. the attach call
+         * will return the base address of the mmid's chunk -
+         * save this base address in a hash table (which maps mmid to base address). */
+        remote_seg = ucs_malloc(sizeof(*remote_seg), "mm_desc");
+        if (remote_seg == NULL) {
+            ucs_fatal("Failed to allocated memory for a remote segment identifier. %m");
+        }
+
+        status = uct_mm_pd_mapper_ops(iface->super.pd)->attach(elem->desc_mmid,
+                                      elem->desc_mpool_size,
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: need to fix indentation

File: src/uct/sm/mm/mm_ep.c [hunk 3]
```diff
@@ -89,6 +111,44 @@ ucs_status_t uct_mm_ep_put_bcopy(uct_ep_h tl_ep, uct_pack_callback_t pack_cb,
...
+                                      elem->desc_chunk_base_addr,
+                                      &remote_seg->address, &remote_seg->cookie);
+        if (status != UCS_OK) {
+            ucs_fatal("Failed to attach to remote mmid:%zu (error=%m)", elem->desc_mmid);
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: %m -> status

**Comment by shamisp** [comment]
Ignore the failure, this is one of the thresholds

File: src/uct/sm/mm/mm_ep.c [hunk 4]
```diff
@@ -109,18 +169,23 @@ static inline ucs_status_t uct_mm_ep_get_remote_elem(uct_mm_ep_t *ep, uint64_t h
...
 
-ucs_status_t uct_mm_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
-                                const void *payload, unsigned length)
+/* A common mm active message sending function.
+ * The first parameter indicates the origin of the call.
+ * 1 - perform am short sending
+ * 0 - perform am bcopy sending
+ */
+static UCS_F_ALWAYS_INLINE ucs_status_t uct_mm_ep_am_common_send(unsigned is_short,
+                                        uct_mm_ep_t *ep, uct_mm_iface_t *iface,
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: indentation

File: src/uct/sm/mm/mm_ep.c [hunk 5]
```diff
@@ -131,30 +196,84 @@ ucs_status_t uct_mm_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
     status = uct_mm_ep_get_remote_elem(ep, head, &elem);
 
     if (status == UCS_OK) {
-        /* write to the remote fifo */
-        *(uint64_t*)(elem + 1) = header;
-        memcpy((void*)(elem + 1) + sizeof(header), payload, length);
-        elem->am_id  = id;
-        elem->flags |= UCT_MM_FIFO_ELEM_FLAG_INLINE;
-        elem->length = length + sizeof(header);
+        if (is_short == 1) {
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: if (is_short)

File: src/uct/sm/mm/mm_ep.c [hunk 6]
```diff
@@ -131,30 +196,84 @@ ucs_status_t uct_mm_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
     status = uct_mm_ep_get_remote_elem(ep, head, &elem);
 
     if (status == UCS_OK) {
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: would be nicer to return if failed

File: src/uct/sm/mm/mm_ep.c (hunk 5 above)
Comment (alinask) [inline src/uct/sm/mm/mm_ep.c]: right. debug leftover...

File: src/uct/sm/mm/mm_ep.c [hunk 7]
```diff
@@ -131,30 +196,84 @@ ucs_status_t uct_mm_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
...
+{
+    uct_mm_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_mm_iface_t);
+    uct_mm_ep_t *ep = ucs_derived_of(tl_ep, uct_mm_ep_t);
+    ucs_status_t status;
+
+    UCT_CHECK_LENGTH(length + sizeof(header),
+                     iface->elem_size - sizeof(uct_mm_fifo_element_t), "am_short");
+
+    status = uct_mm_ep_am_common_send(1, ep, iface, id, length, header, payload,
+                                (void*)ucs_empty_function_return_success, (void*)0x0);
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: indentation
can pass NULL,NULL as last parameters

File: src/uct/sm/mm/mm_ep.c [hunk 8]
```diff
@@ -131,30 +196,84 @@ ucs_status_t uct_mm_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
...
+    if (status == UCS_OK) {
+        ucs_trace_data("MM: AM_SHORT [%p] am_id: %d buf=%p length=%u",
+                        iface, id, payload, length);
     } else {
-        /* couldn't get an available fifo element */
-        return status;
+        ucs_debug("Couldn't get an available FIFO element in am_short");
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: better remove this debug mesage or use ucs_trace_poll

File: src/uct/sm/mm/mm_ep.c [hunk 9]
```diff
@@ -131,30 +196,84 @@ ucs_status_t uct_mm_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
...
+                                uct_pack_callback_t pack_cb, void *arg,
+                                size_t length)
+{
+    uct_mm_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_mm_iface_t);
+    uct_mm_ep_t *ep = ucs_derived_of(tl_ep, uct_mm_ep_t);
+    ucs_status_t status;
+
+    UCT_CHECK_LENGTH(length, iface->config.seg_size, "am_bcopy");
+
+    status = uct_mm_ep_am_common_send(0, ep, iface, id, length, 0, (void*)0x0, pack_cb, arg);
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: (void*)0x0 -> NULL

File: src/uct/sm/mm/mm_ep.c [hunk 10]
```diff
@@ -131,30 +196,84 @@ ucs_status_t uct_mm_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
...
+    if (status == UCS_OK) {
+        ucs_trace_data("MM: AM_BCOPY [%p] am_id: %d buf=%p length=%u",
+                        iface, id, arg, (int)length);
+    } else {
+        ucs_debug("Couldn't get an available FIFO element in am_bcopy");
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: pls remove this debug

File: src/uct/sm/mm/mm_ep.h [hunk 11]
```diff
@@ -63,4 +71,17 @@ ucs_status_t uct_mm_ep_get_bcopy(uct_ep_h ep, uct_unpack_callback_t unpack_cb,
                                  uint64_t remote_addr, uct_rkey_t rkey,
                                  uct_completion_t *comp);
 
+static inline int uct_mm_remote_seg_hash(uct_mm_remote_seg_t *seg)
+{
+    return seg->mmid % UCT_MM_BASE_ADDRESS_HASH_SIZE;
+}
+
+static inline int uct_mm_remote_seg_compare(uct_mm_remote_seg_t *seg1, uct_mm_remote_seg_t *seg2)
```
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.h]: need to return int64_t because mmid is int64_t

File: src/uct/sm/mm/mm_iface.c [hunk 12]
```diff
@@ -152,15 +153,48 @@ static inline void uct_mm_progress_fifo_tail(uct_mm_iface_t *iface)
     iface->recv_fifo_ctl->tail = iface->read_index;
 }
 
+ucs_status_t uct_mm_assign_desc_to_fifo_elem(uct_mm_iface_t *iface,
+                                     uct_mm_fifo_element_t *fifo_elem_p)
```
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.c]: indentation

File: src/uct/sm/mm/mm_iface.c [hunk 13]
```diff
@@ -152,15 +153,48 @@ static inline void uct_mm_progress_fifo_tail(uct_mm_iface_t *iface)
...
     } else {
-        ucs_fatal("bcopy is not implemented");
+        /* read bcopy messages from the receive descriptors */
+        VALGRIND_MAKE_MEM_DEFINED(elem->desc_chunk_base_addr + elem->desc_offset,
+                                  elem->length);
+        status = uct_mm_iface_invoke_am(iface, elem->am_id,
+                                        elem->desc_chunk_base_addr + elem->desc_offset,
+                                        elem->length,
+                                        (uct_mm_recv_desc_t *) (elem->desc_chunk_base_addr +
+                                        elem->desc_offset - iface->rx_headroom) - 1,
```
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.c]: indentation

File: src/uct/sm/mm/mm_iface.c [hunk 14]
```diff
@@ -152,15 +153,48 @@ static inline void uct_mm_progress_fifo_tail(uct_mm_iface_t *iface)
     iface->recv_fifo_ctl->tail = iface->read_index;
 }
 
+ucs_status_t uct_mm_assign_desc_to_fifo_elem(uct_mm_iface_t *iface,
+                                     uct_mm_fifo_element_t *fifo_elem_p)
+{
+    uct_mm_recv_desc_t *desc;
+
+    UCT_TL_IFACE_GET_RX_DESC(&iface->super, iface->recv_desc_mp, desc,
```
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.c]: maybe this function should not return error code if there is no desc, but print fatal error.
another option is to mark FIFO element as "no-skb" which will make the sender stop until there is an skb

File: src/uct/sm/mm/mm_iface.c [hunk 15]
```diff
@@ -207,6 +241,19 @@ void uct_mm_iface_recv_desc_init(uct_iface_h tl_iface, void *obj, uct_mem_h memh
...
+}
+
+static void uct_mm_iface_free_rx_descs(uct_mm_iface_t *iface, unsigned num_elems)
+{
+    uct_mm_fifo_element_t* fifo_elem_p;
+    unsigned i;
+
+    for (i = 0; i < num_elems; i++) {
+        fifo_elem_p = UCT_MM_IFACE_GET_FIFO_ELEM(iface, iface->recv_fifo_elements, i);
+        ucs_mpool_put((uct_mm_recv_desc_t *) (fifo_elem_p->desc_chunk_base_addr +
```
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.c]: maybe add macro to calculate this

File: src/uct/sm/mm/mm_iface.c [hunk 16]
```diff
@@ -301,17 +342,30 @@ static UCS_CLASS_INIT_FUNC(uct_mm_iface_t, uct_pd_h pd, uct_worker_h worker,
...
+    /* initiate the owner bit in all the FIFO elements and assign a receive descriptor
+     * per every FIFO element */
+    for (i = 0; i < mm_config->fifo_size; i++) {
+        fifo_elem_p = UCT_MM_IFACE_GET_FIFO_ELEM(self, self->recv_fifo_elements, i);
+        fifo_elem_p->flags = UCT_MM_FIFO_ELEM_FLAG_OWNER;
+
+        status = uct_mm_assign_desc_to_fifo_elem(self, fifo_elem_p);
+        if (status != UCS_OK) {
+            ucs_error("Failed to allocate a descriptor for MM");
+            goto destroy_descs;
```
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.c]: what if only some descriptors were allocated? uct_mm_iface_free_rx_descs will segfault on the rest

File: src/uct/sm/mm/mm_iface.h [hunk 17]
```diff
@@ -67,10 +69,19 @@ struct uct_mm_iface {
...
-    uint8_t             flags;
-    uint8_t             am_id;     /* active message id */
-    uint16_t            length;    /* length of actual data */
-    /* the data follows here */
+    uint8_t         flags;
+    uint8_t         am_id;          /* active message id */
+    uint16_t        length;         /* length of actual data */
+
+    /* bcopy parameters */
+    size_t          desc_mpool_size;
```
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.h]: maybe rename to desc_size?

File: src/uct/sm/mm/mm_ep.c [hunk 18]
```diff
@@ -109,18 +169,23 @@ static inline ucs_status_t uct_mm_ep_get_remote_elem(uct_mm_ep_t *ep, uint64_t h
...
 }
 
-ucs_status_t uct_mm_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
-                                const void *payload, unsigned length)
+/* A common mm active message sending function.
+ * The first parameter indicates the origin of the call.
+ * 1 - perform am short sending
+ * 0 - perform am bcopy sending
+ */
+static UCS_F_ALWAYS_INLINE ucs_status_t uct_mm_ep_am_common_send(unsigned is_short,
```
Comment (shamisp) [inline src/uct/sm/mm/mm_ep.c]: const unsigned is_short

File: src/uct/sm/mm/mm_iface.c (hunk 16 above)
Comment (alinask) [inline src/uct/sm/mm/mm_iface.c]: it goes to freeing them right after the first one fails which means that all the previous ones were allocated. and it frees up to the failed one. why would it segfault?

File: src/uct/sm/mm/mm_iface.h (hunk 17 above)
Comment (alinask) [inline src/uct/sm/mm/mm_iface.h]: that would imply that the size is of the desc, but it should be of the entire memory pool it belongs to

File: src/uct/sm/mm/mm_ep.c (hunk 9 above)
Comment (shamisp) [inline src/uct/sm/mm/mm_ep.c]: For readability purpose I would define SHORT=1, BCOPY=0

File: src/uct/sm/mm/mm_ep.c (hunk 6 above)
Comment (alinask) [inline src/uct/sm/mm/mm_ep.c]: what do you mean? it returns with a failed status

File: src/uct/sm/mm/mm_ep.c (hunk 10 above)
Comment (alinask) [inline src/uct/sm/mm/mm_ep.c]: why..?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/293/ for details. (Mellanox internal link)

File: src/uct/sm/mm/mm_ep.c (hunk 6 above)
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]:
```
if (status != UCS_OK) {
    retrun status;
}
```

File: src/uct/sm/mm/mm_ep.c (hunk 10 above)
Comment (yosefe) [inline src/uct/sm/mm/mm_ep.c]: because it would be printed all the time, when re-trying to send

File: src/uct/sm/mm/mm_iface.c (hunk 16 above)
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.c]: right, i missed the second parameter - `i`.

File: src/uct/sm/mm/mm_iface.h (hunk 17 above)
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.h]: what do you mean? isn't this the size of the single descriptor?

File: src/uct/sm/mm/mm_ep.c (hunk 6 above)
Comment (alinask) [inline src/uct/sm/mm/mm_ep.c]: this function does return the status in the end in any case, if failed or not

File: src/uct/sm/mm/mm_iface.h (hunk 17 above)
Comment (yosefe) [inline src/uct/sm/mm/mm_iface.h]: ok got it, you need it to attach

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/881/ for details.

**Comment by alinask** [comment]
@shamisp can you please post the failures from your hosts that failed?

**Comment by yosefe** [comment]
looks ok, need to check the failures

**Comment by shamisp** [comment]
It failed on performance thresholds. Please ignore.

**Comment by shamisp** [comment]
:+1:

**Comment by yosefe** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/303/ for details. (Mellanox internal link)

**Pull request merged**